keps/sig-scheduling/5075-dra-consumable-capacity/README.md

```diff
@@ -0,0 +1,1256 @@
+              if err != nil {
+                return false, err
+              }
+              if !consumable {
+                alloc.logger.V(7).Info("Device not consumable", "device", deviceID)
+                continue
+              }
+            }
+            ...
+          }
+        }
+      ...
+    }
+
+    // isSharedDevice checks whether the device is shared.
+    // A device is considered as a shared device if at least one consumableCapacity is defined.
+    func (alloc *allocator) isSharedDevice(slice *draapi.ResourceSlice, deviceIndex int) bool {
+      basicDevice := slice.Spec.Devices[deviceIndex].Basic
+      if basicDevice == nil {
+        return false
+      }
+      return slice.Spec.Devices[deviceIndex].Basic.ConsumableCapacity != nil &&
+        len(slice.Spec.Devices[deviceIndex].Basic.ConsumableCapacity) > 0
+    }
+
+
+    // isConsumable checks whether a device with remaining resources is consumable by the request.
+    func (alloc *allocator) isConsumable(r requestIndices, slice *draapi.ResourceSlice, deviceIndex int) (bool, error) {
+      deviceID := DeviceID{Driver: slice.Spec.Driver, Pool: slice.Spec.Pool.Name, Device: slice.Spec.Devices[deviceIndex].Name}
+      request := &alloc.claimsToAllocate[r.claimIndex].Spec.Devices.Requests[r.requestIndex]
+
+      if IsFullDeviceRequest(*request) {
+        if allocatedShare, found := alloc.allocatedShareCollection[deviceID]; found {
+          if allocatedShare.HasNoShare() {
+            return true, nil
+          }
+          return false, nil
+        }
+        return true, nil
+      }
+
+      consumableCapacity := slice.Spec.Devices[deviceIndex].Basic.ConsumableCapacity
+      var consumable bool
+      var err error
+
+      if allocatedShare, found := alloc.allocatedShareCollection[deviceID]; found {
+        consumable, err = allocatedShare.IsConsumable(request.Resources.Requests, consumableCapacity)
+      } else {
+        consumable, err = NewAllocatedShare().IsConsumable(request.Resources.Requests, consumableCapacity)
+      }
+      return consumable, err
+    }
+    ```
+
+- add request's `resources` in `internalDeviceResult` for updating ResourceClaim status correspondingly
+
+  ```go
+  type internalDeviceResult struct {
+    request     string
+    id          DeviceID
+    slice       *draapi.ResourceSlice
+    resources   *map[resourceapi.QualifiedName]resource.Quantity
+    adminAccess *bool
+  }
+  ```
+
+  ```go
+  func (a *Allocator) Allocate(ctx context.Context, node *v1.Node) (finalResult []resourceapi.AllocationResult, finalErr error) {
+        ...
+      for i, internal := range internalResult.devices {
+        allocationResult.Devices.Results[i] = resourceapi.DeviceRequestAllocationResult{
+          ...
+          AllocatedShare: internal.resources,
+        }
+      }
+  }
+  ```
+
+### Test Plan
```

Comment:
this part is missing

Reply:
Looks good and very thorough (thanks!) to me.